edx_api/enrollments/__init__.py

```diff
@@ -22,6 +25,88 @@ def __init__(self, requester, base_url):
         """
         self.requester = requester
         self.base_url = base_url
+        self._course_details = {}
+
+    def _get_course_details(self, course_id):
+        """
+        Retrieves course details.
+
+        Args:
+            course_id (str): Course id.
+        """
+        if course_id in self._course_details:
+            return self._course_details[course_id]
+
+        req_url = urljoin(self.base_url, self.course_details_url.format(course_key=course_id))
+        resp = self.requester.get(req_url)
+        resp.raise_for_status()
+        resp_json = resp.json()
+        self._course_details[course_id] = resp_json
+        return resp_json
+
+    def _parse_enrollments(self, enrollments_list):
+        """
+        Parse enrollments response into :class:`Enrollments`.
+        """
+        def populate_course_details(enrollment):
+            """Populates course details if missing."""
+            enrollment.pop('previous', None)
+            enrollment.pop('next', None)
+            if enrollment.get('course_details'):
+                return enrollment
+            course_details = self._get_course_details(enrollment['course_id'])
+            course_details['course_start'] = course_details['start']
+            course_details['course_end'] = course_details['end']
+            enrollment['course_details'] = course_details
+            return enrollment
+
+        enrollments = [populate_course_details(enrl) for enrl in enrollments_list]
+        return Enrollments(enrollments)
+
+    def _get_list_enrollments(self, params=None):
+        """
+        Private method to centralize request code.
+        """
+        req_url = urljoin(self.base_url, self.enrollment_list_url)
+        resp = self.requester.get(req_url, params=params)
+        resp.raise_for_status()
+        resp_json = resp.json()
+        next_url = resp_json.get('next')
+        results = resp_json['results']
+
+        while next_url:
+            prev_url = urlparse(resp_json['next'])
+            qstr = parse_qs(prev_url.query)
+            try:
+                cursor = qstr['cursor'][0]
+            except (KeyError, IndexError):
```

Comment:
Could we use some conditionals instead of exceptions here?
There are some reasonable ways to check if a key exists in a dictionary.
```
if 'cursor' in qstr and len(qstr.get('cursor')) != 0:
```
I think this kind of exception handling as conditionals is a sort of anti-pattern, but maybe that's just me.

Reply:
I'm OK with either one. Usually people prefer EAFP. I'll change this.